aspnetcore/razor-components/forms-validation.md

```diff
@@ -0,0 +1,181 @@
+
+    [Required]
+    public string Classification { get; set; }
+
+    [Range(1, 100000, ErrorMessage = "Accommodation invalid (1-100000).")]
+    public int MaximumAccommodation { get; set; }
+
+    [Required]
+    [Range(typeof(bool), "true", "true", 
+        ErrorMessage = "Form disallowed for unapproved ships.")]
+    public bool IsValidatedDesign { get; set; }
+
+    [Required]
+    public DateTime ProductionDate { get; set; }
+}
+```
+
+The following Starship Database form validates user input using the validation defined in the `Starship` model:
+
+```cshtml
+@page "/FormsValidation"
+
+<h1>Starfleet Starship Database</h1>
+
+<h2>New Ship Entry Form</h2>
+
+<EditForm Model="@starship" OnValidSubmit="@HandleValidSubmit">
+    <DataAnnotationsValidator />
+    <ValidationSummary />
+
+    <p>
+        <label for="identifier">Identifier: </label>
+        <InputText id="identifier" bind-Value="@starship.Identifier" />
+    </p>
+    <p>
+        <label for="description">Description (optional): </label>
+        <InputTextArea Id="description" bind-Value="@starship.Description" />
+    </p>
+    <p>
+        <label for="classification">Primary Classification: </label>
+        <InputSelect id="classification" bind-Value="@starship.Classification">
+            <option value"">Select classification ...</option>
+            <option value="Defense">Defense</option>
+            <option value="Exploration">Exploration</option>
+            <option value="Diplomacy">Diplomacy</option>
+        </InputSelect>
+    </p>
+    <p>
+        <label for="accommodation">Maximum Accommodation: </label>
+        <InputNumber id="accommodation" 
+            bind-Value="@starship.MaximumAccommodation" />
+    </p>
+    <p>
+        <label for="valid">Engineering Approval: </label>
+        <InputCheckbox id="valid" bind-Value="@starship.IsValidatedDesign" />
+    </p>
+    <p>
+        <label for="productionDate">Production Date: </label>
+        <InputDate Id="productionDate" bind-Value="@starship.ProductionDate" />
+    </p>
+
+    <button type="submit">Submit</button>
+
+    <p>
+        <a href="http://www.startrek.com/">Star Trek</a>, 
+        &copy;1966-2019 CBS Studios, Inc. and 
+        <a href="http://www.paramount.com">Paramount Pictures</a>
+    </p>
+</EditForm>
+
+@functions {
+    private Starship starship = new Starship();
+
+    void HandleValidSubmit()
+    {
+        Console.WriteLine("OnValidSubmit");
+    }
+}
+```
+
+The `<EditForm>` creates an `EditContext` as a [cascading value](xref:razor-components/components#cascading-values-and-parameters) that tracks metadata about the edit process, including what fields have been modified and the current validation messages. The `<EditForm>` also provides convenient events for valid and invalid submits (`OnValidSubmit`, `OnInvalidSubmit`). Alternatively, you can use `OnSubmit` directly if you want to trigger the validation and check field values yourself.
```

Comment:
Alternatively, use `OnSubmit` to trigger the validation and check field values with user supplied validation code.

Or something like that.